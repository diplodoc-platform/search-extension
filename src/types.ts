import type {ISearchWorkerConfig} from '@diplodoc/client';

export enum Confidence {
    Phrased = 'phrased',
    Sparsed = 'sparsed',
}

export interface WorkerConfig extends ISearchWorkerConfig {
    tolerance: number;
    confidence: `${Confidence}` | Confidence;
    resources: {
        index: string;
        registry: string;
        language?: string;
    };
}

export type Position = [number, number];

export type IndexDocument = {
    url: string;
    title: string;
    content: string;
    breadcrumbs?: string[];
    keywords?: string[];
};

export type Registry = Record<string, IndexDocument>;
