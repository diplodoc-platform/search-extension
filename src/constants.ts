export const MERGE_TOLERANCE = 5;

export const INDEX_FIELDS = {
    title: 10,
    keywords: 8,
    content: 1,
};
