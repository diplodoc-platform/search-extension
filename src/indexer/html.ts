// eslint-disable-next-line @typescript-eslint/no-redeclare
import type {HTMLElement, Node, TextNode} from 'node-html-parser';

import {NodeType, parse} from 'node-html-parser';

function isElementNode(node: Node): node is HTMLElement {
    return node.nodeType === NodeType.ELEMENT_NODE;
}

function isTextNode(node: Node): node is TextNode {
    return node.nodeType === NodeType.TEXT_NODE;
}

function isBlockTag(tag: string) {
    return 'div|p|li|td|section|br'.includes(tag);
}

export function html2text(html: string) {
    const node = parse(html, {
        lowerCaseTagName: true,
        blockTextElements: {
            script: false,
            noscript: false,
            style: false,
            pre: true,
        },
    });

    let currentBlock: string[] = [];
    let prependWhitespace = false;

    const blocks = [currentBlock];
    const work: (Node | null)[] = [node];

    while (work.length) {
        const node = work.shift();

        if (!node) {
            if (currentBlock.length) {
                blocks.push((currentBlock = []));
                prependWhitespace = false;
            }
            continue;
        }

        if (isElementNode(node)) {
            const tagName = (node.tagName || '').toLowerCase();

            if (node.attributes['data-no-index']) {
                continue;
            }

            if (tagName === 'pre') {
                currentBlock.push(html2text(node.rawText));
            } else if (isBlockTag(tagName)) {
                work.unshift(null, ...node.childNodes, null);
            } else {
                work.unshift(...node.childNodes);
            }
        }

        if (isTextNode(node)) {
            if (node.isWhitespace) {
                // Whitespace node, postponed output
                prependWhitespace = true;
            } else {
                let text = node.text;
                if (prependWhitespace) {
                    text = ' ' + text;
                    prependWhitespace = false;
                }
                currentBlock.push(text);
            }
        }
    }

    // Normalize each line's whitespace
    return blocks
        .map((block) =>
            block
                .join(' ')
                .trim()
                .replace(/\s{2,}/g, ' ')
                .replace(/\s([.,!?:;])/g, '$1'),
        )
        .join('\n')
        .trimEnd();
}
