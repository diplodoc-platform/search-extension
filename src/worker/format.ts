import type {SearchSuggestPageItem} from '@diplodoc/components';
import type {Position, Registry, WorkerConfig} from '../types';
import type {SearchResult} from './search';
import type {Score} from './score';

const MAX_LENGTH = 200;
const SHORT_HEAD = 20;

type Trimmer = (text: string, score: Score) => [string, Position[]];

export function format(
    {base, mark}: WorkerConfig,
    result: SearchResult[],
    registry: Registry,
    trim: Trimmer,
): SearchSuggestPageItem[] {
    return result.map((entry) => {
        const doc = registry[entry.ref];
        const item = {
            type: 'page',
            link: `${base}/${entry.ref}`,
            title: doc.title,
            description: doc.content.slice(0, MAX_LENGTH),
        } as SearchSuggestPageItem;

        const fields = entry.scores;

        for (const [field, score] of Object.entries(fields)) {
            switch (field) {
                case 'title':
                    item.title = highlighte(mark, doc.title, score.positions);
                    break;
                case 'content':
                    item.description = highlighte(mark, ...trim(doc.content, score));
                    break;
            }
        }

        return item;
    });
}

export function short(text: string, score: Score): [string, Position[]] {
    const {positions, maxScorePosition: position} = score;
    const [before, content, after] = split(text, position);
    const head = before.length > SHORT_HEAD ? '...' + before.slice(-SHORT_HEAD) : before;
    const tail = after.slice(0, Math.max(0, MAX_LENGTH - head.length - content.length));

    return [head + content + tail, remap(position[0] - head.length, positions)];
}

export function long(text: string, score: Score): [string, Position[]] {
    const {positions} = score;

    if (text.length <= MAX_LENGTH) {
        return [text, positions];
    }

    const [before, content, after] = split(text, positions[0]);

    if (content.length >= MAX_LENGTH) {
        return [content.slice(0, MAX_LENGTH), [[0, MAX_LENGTH]]];
    }

    let head = 0;
    let result = content;

    const befores = before.split('\n');
    const afters = after.split('\n');

    let action = prepend;
    while (result.length < MAX_LENGTH) {
        action = action();
    }

    const length = result.length;
    result = result.trimStart();
    head -= length - result.length;

    return [result, remap(head, positions)];

    function prepend() {
        if (befores.length) {
            const part = befores.pop() as string;
            const overage = Math.max(part.length - MAX_LENGTH, 0);

            head += part.length - overage + 1;
            result = '\n' + part.slice(overage, part.length) + result;
        }

        return append;
    }

    function append() {
        if (afters.length) {
            const part = afters.shift() as string;
            const overage = Math.max(result.length + part.length - MAX_LENGTH, 0);

            result = result + part.slice(0, part.length - overage) + '\n';
        }

        return prepend;
    }
}

function highlighte(mark: string, text: string, positions: Position[]) {
    return positions.reduceRight((result, position) => {
        const [before, content, after] = split(result, position);

        return `${before}<span class="${mark}">${content}</span>${after}`;
    }, text);
}

function split(text: string, position: Position) {
    const before = text.slice(0, position[0]);
    const content = text.slice(position[0], position[1]);
    const after = text.slice(position[1]);

    return [before, content, after];
}

function remap(dl: number, positions: Position[]): Position[] {
    return positions
        .map(([start, end]) => [start - dl, end - dl])
        .filter(([start]) => start >= 0) as [number, number][];
}
