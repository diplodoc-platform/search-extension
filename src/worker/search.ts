import type {Index} from 'lunr';
import type {Score} from './score';
import type {WorkerConfig} from '../types';

// @ts-ignore
import {Query, QueryParser} from 'lunr';

import {INDEX_FIELDS} from '../constants';

import {phrased, sparsed} from './score';

const isStrategy = (candidate: unknown): candidate is Index.QueryBuilder =>
    typeof candidate === 'function';

const withIndex = (index: Index) => (builder: Index.QueryBuilder) =>
    function withIndex() {
        return index.query(builder);
    };

type Strategy = () => Index.Result[];

interface FixedIndex extends Index {
    fields: string[];
}

interface FixedClause extends Query.Clause {
    presence: Query.presence;
}

const makeStrategies = (tolerance: number, index: Index, clauses: FixedClause[], sealed: boolean) =>
    [
        tolerance >= 0 &&
            function precise(query: Query) {
                query.clauses = copy(clauses);
            },
        tolerance >= 0 &&
            !sealed &&
            function preciseUnsealed(query: Query) {
                query.clauses = copy(clauses);

                for (let i = query.clauses.length - 1; i >= 0; i--) {
                    const clause = query.clauses[i] as FixedClause;
                    if (clause.presence !== Query.presence.PROHIBITED) {
                        query.clauses[i] = wildcard(clause, Query.wildcard.TRAILING);
                        break;
                    }
                }
            },
        tolerance >= 1 &&
            function trailingWildcard(query: Query) {
                query.clauses = copy(clauses).map((clause) => {
                    if (clause.presence !== Query.presence.PROHIBITED) {
                        return wildcard(clause, Query.wildcard.TRAILING);
                    }
                    return clause;
                });
            },
        tolerance >= 2 &&
            function bothWildcard(query: Query) {
                query.clauses = copy(clauses).map((clause) => {
                    if (clause.presence !== Query.presence.PROHIBITED) {
                        // eslint-disable-next-line no-bitwise
                        return wildcard(clause, Query.wildcard.LEADING | Query.wildcard.TRAILING);
                    }
                    return clause;
                });
            },
    ]
        .filter(isStrategy)
        .map(withIndex(index));

export type SearchResult = Index.Result & {scores: Record<string, Score>};

export function search(
    {tolerance, confidence}: Pick<WorkerConfig, 'confidence' | 'tolerance'>,
    index: Index,
    query: string,
    count: number,
    sealed = true,
) {
    const parser = new QueryParser(query, new Query((index as FixedIndex).fields));
    const clauses = parser.parse().clauses as FixedClause[];
    const terms = clauses.map((clause) => clause.term);
    const strategies = makeStrategies(tolerance, index, clauses, sealed);
    const refs = new Set<string>();

    const score = confidence === 'sparsed' ? sparsed : phrased;
    const results: SearchResult[] = [];
    while (refs.size < count && strategies.length) {
        const strategy = strategies.shift() as Strategy;
        const match = strategy();

        for (const entry of match) {
            if (!refs.has(entry.ref)) {
                refs.add(entry.ref);

                results.push({
                    ...entry,
                    scores: score(entry, terms),
                });
            }
        }
    }

    return results.sort(byMaxScore).slice(0, count);
}

function wildcard(clause: FixedClause, mode: Query.wildcard) {
    const result = {...clause};

    const requiredLength =
        [
            // eslint-disable-next-line no-bitwise
            mode & Query.wildcard.TRAILING ? 2 : 0,
            // eslint-disable-next-line no-bitwise
            mode & Query.wildcard.LEADING ? 2 : 0,
        ].reduce((a, b) => a + b, 0) + 1;

    if (result.term.length < requiredLength) {
        return result;
    }

    // eslint-disable-next-line no-bitwise
    if (mode & Query.wildcard.TRAILING) {
        result.term = result.term + '*';
    }

    // eslint-disable-next-line no-bitwise
    if (mode & Query.wildcard.LEADING) {
        result.term = '*' + result.term;
    }

    result.wildcard = mode;
    result.usePipeline = false;

    return result;
}

function byMaxScore(a: SearchResult, b: SearchResult) {
    const aScore = getMaxScore(a);
    const bScore = getMaxScore(b);

    return bScore - aScore;
}

function getMaxScore(result: SearchResult) {
    let score = 0;
    for (const [field] of Object.entries(INDEX_FIELDS)) {
        const scores = result.scores[field];
        if (scores) {
            score = Math.max(scores.score, score);
        }
    }

    return score;
}

function copy(clauses: FixedClause[]) {
    return clauses.slice().map((clause) => ({...clause}));
}
