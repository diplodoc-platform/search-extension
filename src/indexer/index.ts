import type {DocPageData} from '@diplodoc/components';

import lunr, {Builder} from 'lunr';

import {INDEX_FIELDS} from '../constants';

import {langs} from './langs';
import {html2text} from './html';

type DocumentInfo = {
    title: string;
    content: string;
    keywords: string[];
};

export enum ReleaseFormat {
    JSONP = 'jsonp',
    RAW = 'raw',
}

export class Indexer {
    private indices: Record<string, Builder> = {};

    private docs: Record<string, Record<string, DocumentInfo>> = {};

    /**
     * Stores list of current non empty language indexes.
     */
    get langs() {
        return Object.keys(this.indices);
    }

    /**
     * Adds new entry to search index
     *
     * @param lang - index language
     * @param url - document url, used as index uniq id
     * @param data - document json data
     *
     * @returns {void}
     */
    add(
        lang: string,
        url: string,
        data: Pick<DocPageData, 'title' | 'html' | 'meta' | 'leading' | 'toc'>,
    ) {
        if (!this.indices[lang]) {
            this.init(lang);
        }

        const {leading, toc, meta = {}} = data;

        // @ts-ignore
        if (leading || meta.noindex || meta.noIndex) {
            return;
        }

        const content = html2text(data.html);
        const title = meta.title || data.title || toc.title || '';
        const keywords = meta.keywords || [];

        this.docs[lang][url] = {title, content, keywords};
        this.indices[lang].add({
            ...this.docs[lang][url],
            url,
        });
    }

    /**
     * Dumps index and registry for target language.
     *
     * @param lang - index language
     * @param format - output format
     *
     * @returns {{index: Index, registry: Registry}}
     */
    release(lang: string, format = ReleaseFormat.JSONP) {
        const index = this.indices[lang].build();
        const registry = this.docs[lang];

        if (format === ReleaseFormat.JSONP) {
            return {
                index: 'self.index=' + JSON.stringify(index),
                registry: 'self.registry=' + JSON.stringify(registry),
            };
        }

        return {index, registry};
    }

    private init(lang: string) {
        const index = new Builder();

        if (langs[lang]) {
            langs[lang](lunr);
            // @ts-ignore
            index.use(lunr.multiLanguage('en', lang));
        }

        index.ref('url');

        for (const [field, boost] of Object.entries(INDEX_FIELDS)) {
            index.field(field, {boost});
        }

        index.metadataWhitelist = ['position'];

        this.indices[lang] = index;
        this.docs[lang] = {};
    }
}
