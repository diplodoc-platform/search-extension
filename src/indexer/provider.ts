import type {BuildRun, EntryInfo, SearchProvider} from '@diplodoc/cli';
import type {WorkerConfig} from '../worker';

import {extname, join} from 'node:path';
import {createHash} from 'node:crypto';
import lunr, {Builder} from 'lunr';

import {INDEX_FIELDS} from '../constants';

import {html2text} from './html';
import {langs} from './langs';

export type ProviderConfig = Pick<WorkerConfig, 'tolerance' | 'confidence'> & {
    enabled: boolean;
};

export enum ReleaseFormat {
    JSONP = 'jsonp',
    RAW = 'raw',
}

type DocumentInfo = {
    title: string;
    content: string;
    keywords: string[];
};

export class Indexer {
    private indices: Record<string, Builder> = {};

    private docs: Record<string, Record<string, DocumentInfo>> = {};

    /**
     * Stores list of current non empty language indexes.
     */
    get langs() {
        return Object.keys(this.indices);
    }

    /**
     * Adds new entry to search index
     *
     * @param lang - index language
     * @param url - document url, used as index uniq id
     * @param data - document json data
     *
     * @returns {void}
     */
    add(lang: string, url: string, data: Pick<EntryInfo, 'title' | 'html' | 'meta' | 'leading'>) {
        if (!this.indices[lang]) {
            this.init(lang);
        }

        const {leading, title = '', meta = {}} = data;

        // @ts-ignore
        if (leading || meta.noindex || meta.noIndex) {
            return;
        }

        const content = html2text(data.html || '');
        const keywords = meta.keywords || [];

        this.docs[lang][url] = {title, content, keywords};
        this.indices[lang].add({
            ...this.docs[lang][url],
            url,
        });
    }

    /**
     * Dumps index and registry for target language.
     *
     * @param lang - index language
     * @param format - output format
     *
     * @returns {{index: Index, registry: Registry}}
     */
    release(lang: string, format = ReleaseFormat.JSONP) {
        const index = this.indices[lang].build();
        const registry = this.docs[lang];

        if (format === ReleaseFormat.JSONP) {
            return {
                index: 'self.index=' + JSON.stringify(index),
                registry: 'self.registry=' + JSON.stringify(registry),
            };
        }

        return {index, registry};
    }

    private init(lang: string) {
        const index = new Builder();

        if (langs[lang]) {
            langs[lang](lunr);
            // @ts-ignore
            index.use(lunr.multiLanguage('en', lang));
        }

        index.ref('url');

        for (const [field, boost] of Object.entries(INDEX_FIELDS)) {
            index.field(field, {boost});
        }

        index.metadataWhitelist = ['position'];

        this.indices[lang] = index;
        this.docs[lang] = {};
    }
}

export class LocalSearchProvider implements SearchProvider {
    private run: BuildRun;

    private _config: ProviderConfig;

    private indexer: Indexer;

    private outputDir: string;

    private apiLink: string;

    private nocache: string;

    constructor(run: BuildRun, config: ProviderConfig) {
        this.run = run;
        this._config = config;
        this.indexer = new Indexer();

        this.outputDir = '_search';
        this.apiLink = join(this.outputDir, 'api.js');
        this.nocache = String(Date.now());
    }

    async add(path: string, lang: string, info: EntryInfo) {
        if (!info.html) {
            return;
        }

        const url = path.replace(extname(path), '') + '.html';

        this.indexer.add(lang, url, info);
    }

    async release() {
        await this.run.copy(
            join(this.run.assetsPath, 'search-extension', 'api.js'),
            join(this.run.output, this.apiLink),
        );

        for (const lang of this.indexer.langs) {
            const {index, registry} = await this.indexer.release(lang);

            const indexHash = hash(index as string);
            const registryHash = hash(registry as string);
            const indexLink = this.indexLink(lang, indexHash);
            const registryLink = this.registryLink(lang, registryHash);
            const languageLink = this.languageLink(lang);
            const resourcesLink = this.resourcesLink(lang);
            const pageLink = this.pageLink(lang);

            await this.run.write(join(this.run.output, indexLink), index as string);
            await this.run.write(join(this.run.output, registryLink), registry as string);
            await this.run.write(
                join(this.run.output, resourcesLink),
                this.resources(indexLink, registryLink, languageLink),
            );
            await this.run.write(join(this.run.output, pageLink), await this.run.search.page(lang));

            if (languageLink) {
                await this.run.copy(
                    join(this.run.assetsPath, 'search-extension', 'langs', lang + '.js'),
                    join(this.run.output, languageLink),
                );
            }
        }
    }

    config(lang: string) {
        return {
            ...this._config,
            api: this.apiLink,
            provider: 'local',
            link: this.pageLink(lang),
        };
    }

    resourcesLink(lang: string) {
        return join(this.outputDir, lang, `${this.nocache}-resources.js`);
    }

    private resources(indexLink: string, registryLink: string, languageLink: string) {
        const resources = {
            index: indexLink,
            registry: registryLink,
            language: languageLink || undefined,
        };

        return `window.__DATA__.search.resources = ${JSON.stringify(resources)};`;
    }

    private indexLink(lang: string, hash: string) {
        return join(this.outputDir, lang, `${hash}-index.js`);
    }

    private registryLink(lang: string, hash: string) {
        return join(this.outputDir, lang, `${hash}-registry.js`);
    }

    private languageLink(lang: string) {
        if (!langs.includes(lang)) {
            return '';
        }

        return join(this.outputDir, lang, `language.js`);
    }

    private pageLink(lang: string) {
        return join(this.outputDir, lang, `index.html`);
    }
}

function hash(content: string) {
    const hash = createHash('sha256');

    hash.update(content);

    return hash.digest('hex').slice(0, 12);
}
