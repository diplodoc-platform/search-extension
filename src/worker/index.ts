/// <reference no-default-lib="true"/>
/// <reference lib="ES2019" />
/// <reference lib="webworker" />

/* eslint-disable new-cap */
import type {Registry, WorkerConfig} from '../types';
import type {ISearchWorkerApi} from '@diplodoc/client';
import type {Builder} from 'lunr';
import type {SearchResult} from './search';

import lunr, {Index} from 'lunr';

import {search} from './search';
import {format, long, paginateResult, short} from './format';

export {WorkerConfig};

// Default type of `self` is `WorkerGlobalScope & typeof globalThis`
// https://github.com/microsoft/TypeScript/issues/14877
declare const self: ServiceWorkerGlobalScope & {
    config?: WorkerConfig;
    api?: ISearchWorkerApi;
    language?: (lunr: unknown) => Builder.Plugin;
    index: object;
    registry: Registry;
};

const NOT_INITIALIZED = {
    message: 'Worker is not initialized with required config!',
    code: 'NOT_INITIALIZED',
};

const MAX_COUNT_RESULT = 100;

export function AssertConfig(config: unknown): asserts config is WorkerConfig {
    if (!config) {
        throw NOT_INITIALIZED;
    }
}

let config: WorkerConfig | null = null;
let index: Index | null = null;
let registry: Registry | null = null;

let lastQuery: string | null = null;
let lastResult: SearchResult[] | null = null;

self.api = {
    async init() {
        config = {
            tolerance: 2,
            confidence: 'phrased',
            ...self.config,
        } as WorkerConfig;
    },

    async suggest(query, count) {
        AssertConfig(config);

        const [index, registry] = await load();
        const results = search(config, index, query, count, false);

        return format(config, results, registry, short);
    },

    async search(query, count, page) {
        AssertConfig(config);

        const [index, registry] = await load();

        let result: SearchResult[];

        if (lastQuery === query && lastResult) {
            result = lastResult;
        } else {
            result = search(config, index, query, MAX_COUNT_RESULT, true);

            lastQuery = query;
            lastResult = result;
        }

        const {items, total} = paginateResult(result, count, page);

        return {
            items: format(config, items, registry, long),
            total,
        };
    },
} as ISearchWorkerApi;

async function load(): Promise<[Index, Registry]> {
    AssertConfig(config);

    if (index && registry) {
        return [index, registry];
    }

    const promise: Promise<[Index, Registry]> = (async () => {
        const scripts = [
            `${config.base}/${config.resources.index}`,
            `${config.base}/${config.resources.registry}`,
            config.resources.language && `${config.base}/${config.resources.language}`,
        ].filter(Boolean) as string[];

        // Load resources using importScripts instead of fetch
        // because fetch will produce CORS errors on file:// protocol
        importScripts(...scripts);

        const {index, registry, language} = self;

        if (language) {
            language(lunr);
        }

        return [Index.load(index), registry];
    })();

    promise.catch((error) => {
        // eslint-disable-next-line no-console
        console.error(error);
    });

    [index, registry] = await promise;

    return [index, registry];
}
