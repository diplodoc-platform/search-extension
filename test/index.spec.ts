import type {Index} from 'lunr';
import type {Registry, WorkerConfig} from '../src/types';
import type {SearchSuggestPageItem} from '@diplodoc/components';

import {Indexer, ReleaseFormat} from '../src/indexer';
import {search} from '../src/worker/search';
import {format, short} from '../src/worker/format';

const Lorem = [
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    'Integer sit amet enim velit.',
    'Nam facilisis eget magna non blandit.',
    'Sed semper, dui ut suscipit semper, nibh justo tempor purus, quis placerat enim dolor vitae neque.',
    'Vivamus dignissim nunc et tortor vulputate maximus.',
    'Fusce lobortis pretium lectus, non pretium mi rhoncus quis.',
    'Curabitur blandit imperdiet metus id luctus.',
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    'Aenean lobortis ligula a mauris posuere, luctus pretium mauris ultrices.',
];

const LongLorem = 'Lorem ipsum dolor '.repeat(100);

const Code = 'crm.stagehistory.list';

const item = ({link, title, description}: SearchSuggestPageItem) => `
    <a href="${link}">
        <div>${title}</div>
        <div>${description}</div>
    </a>
`;

describe('suggest', () => {
    const lang = 'ru';
    let indexer: Indexer;
    let uid = 1;

    function suggest(query: string, config: Pick<WorkerConfig, 'confidence' | 'tolerance'>) {
        const {index, registry} = indexer.release(lang, ReleaseFormat.RAW);

        const results = search(config, index as Index, query, 10, false);

        return format({base: './', mark: 'mark'}, results, registry as Registry, short).map(item);
    }

    function add(html: string, title = '') {
        indexer.add(lang, String(uid++), {
            html,
            title,
            leading: false,
            meta: {},
        });
    }

    beforeEach(() => {
        indexer = new Indexer();
    });

    it('should match html content', () => {
        add(Lorem.slice(0, 2).join(' '));
        add(Lorem.slice(1, 3).join(' '));

        const config = {confidence: 'phrased', tolerance: 2} as const;

        expect(suggest('Lorem ipsum', config)).toMatchSnapshot();
    });

    it('should match title content', () => {
        add(Lorem.slice(1, 3).join(' '), 'Lorem ipsum 1');
        add(Lorem.slice(2, 4).join(' '), 'Lorem ipsum 2');

        const config = {confidence: 'phrased', tolerance: 2} as const;

        expect(suggest('Lorem ipsum', config)).toMatchSnapshot();
    });

    it('should score longest phrase', () => {
        add(Lorem.slice(0, 3).join(' '));
        add(Lorem.slice(1, 5).join(' '));

        const config = {confidence: 'phrased', tolerance: 2} as const;

        expect(suggest('enim dolor vitae', config)).toMatchSnapshot();
    });

    it('should match code', () => {
        add(Code);

        const config = {confidence: 'phrased', tolerance: 2} as const;

        expect(suggest('stagehistory', config)).toMatchSnapshot();
    });

    it('should format very long result', () => {
        add(LongLorem);

        const config = {confidence: 'phrased', tolerance: 0} as const;

        expect(suggest('Lorem ipsum', config)).toMatchSnapshot();
    });
});
