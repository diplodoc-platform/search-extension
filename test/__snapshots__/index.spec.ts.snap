// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suggest should match code 1`] = `
Array [
  "
    <a href=\\"./7\\">
        <div></div>
        <div><span class=\\"mark\\">crm.stagehistory.list</span></div>
    </a>
",
]
`;

exports[`suggest should match html content 1`] = `
Array [
  "
    <a href=\\"./1\\">
        <div></div>
        <div><span class=\\"mark\\">Lorem ipsum</span> dolor sit amet, consectetur adipiscing elit. Integer sit amet enim velit.</div>
    </a>
",
]
`;

exports[`suggest should match title content 1`] = `
Array [
  "
    <a href=\\"./3\\">
        <div><span class=\\"mark\\">Lorem ipsum</span> 1</div>
        <div>Integer sit amet enim velit. Nam facilisis eget magna non blandit.</div>
    </a>
",
  "
    <a href=\\"./4\\">
        <div><span class=\\"mark\\">Lorem ipsum</span> 2</div>
        <div>Nam facilisis eget magna non blandit. Sed semper, dui ut suscipit semper, nibh justo tempor purus, quis placerat enim dolor vitae neque.</div>
    </a>
",
]
`;

exports[`suggest should score longest phrase 1`] = `
Array [
  "
    <a href=\\"./6\\">
        <div></div>
        <div>...urus, quis placerat <span class=\\"mark\\">enim dolor vitae</span> neque. Vivamus dignissim nunc et tortor vulputate maximus.</div>
    </a>
",
  "
    <a href=\\"./5\\">
        <div></div>
        <div>Lorem ipsum <span class=\\"mark\\">dolor</span> sit amet, consectetur adipiscing elit. Integer sit amet enim velit. Nam facilisis eget magna non blandit.</div>
    </a>
",
]
`;
